plugins/services/src/js/components/modals/NewCreateServiceModal.js

```diff
@@ -1,100 +1,237 @@
+  'handleConfirmGoBack',
   'handleConvertToPod',
+  'handleGoBack',
   'handleJSONToggle',
+  'handleRouterWillLeave',
   'handleServiceChange',
   'handleServiceErrorChange',
   'handleServiceReview',
   'handleServiceRun',
   'handleServiceSelection',
-  'handleTabChange'
+  'handleStoreChange',
+  'handleTabChange',
+  'onMarathonStoreServiceCreateError',
+  'onMarathonStoreServiceCreateSuccess',
+  'onMarathonStoreServiceEditError',
+  'onMarathonStoreServiceEditSuccess'
 ];
 
-class NewServiceFormModal extends Component {
+class NewCreateServiceModal extends Component {
   constructor() {
     super(...arguments);
 
-    this.state = this.getResetState();
+    const {location, params, route} = this.props;
+    const {router} = this.context;
+    const isEdit = this.isLocationEdit(location);
+    const serviceID = decodeURIComponent(params.id || '/');
+    const service = isEdit ? DCOSStore.serviceTree.findItemById(serviceID) : null;
+    const isSpecificVersion = service instanceof Application && params.version;
+    let serviceConfig = new Application(
+      Object.assign({id: getBaseID(serviceID)}, DEFAULT_APP_SPEC)
+    );
+
+    if (isEdit && service instanceof Service && !isSpecificVersion) {
+      serviceConfig = service.getSpec();
+    }
+
+    if (isEdit && isSpecificVersion) {
+      serviceConfig = service.getVersions().get(params.version);
+    }
+
+    this.state = {
+      activeTab: null,
+      errors: null,
+      hasChangesApplied: false,
+      isConfirmOpen: false,
+      isJSONModeActive: false,
+      isOpen: true,
+      isPending: false,
+      service,
+      serviceConfig,
+      serviceFormActive: isEdit, // Switch directly to form/json if edit
+      serviceFormHasErrors: false,
+      serviceJsonActive: false,
+      servicePickerActive: !isEdit, // Switch directly to form/json if edit
+      serviceReviewActive: false
+    };
 
     METHODS_TO_BIND.forEach((method) => {
       this[method] = this[method].bind(this);
     });
-  }
 
-  /**
-   * @override
-   */
-  componentWillUpdate(nextProps) {
-    const requestCompleted = this.props.isPending && !nextProps.isPending;
-    const shouldClose = requestCompleted && !nextProps.errors;
+    // Add store change listeners the traditional way as React Router is
+    // not able to pass down correct props if we are using StoreMixin
+    MarathonStore.addChangeListener(
+      MARATHON_SERVICE_CREATE_ERROR,
+      this.onMarathonStoreServiceCreateError
+    );
+    MarathonStore.addChangeListener(
+      MARATHON_SERVICE_CREATE_SUCCESS,
+      this.onMarathonStoreServiceCreateSuccess
+    );
+    MarathonStore.addChangeListener(
+      MARATHON_SERVICE_EDIT_ERROR,
+      this.onMarathonStoreServiceEditError
+    );
+    MarathonStore.addChangeListener(
+      MARATHON_SERVICE_EDIT_SUCCESS,
+      this.onMarathonStoreServiceEditSuccess
+    );
 
-    if (shouldClose) {
-      this.handleClose();
+    router.setRouteLeaveHook(route, this.handleRouterWillLeave);
```

Comment:
Hey @mlunoe , that code snippet looked a bit weird, since it never de-registers the hook, so I started digging into the react-router internals. And found that this:

* https://github.com/ReactTraining/react-router/blob/6eeb7ad358f987520f5b519e48bdd31f725cbade/modules/RouterUtils.js#L4
* leading to this https://github.com/ReactTraining/react-router/blob/6eeb7ad358f987520f5b519e48bdd31f725cbade/modules/createTransitionManager.js#L188

That looks it returns a function that should be called to unregister the callback. So, to be correct, you should do something like:

```js
componentDidMount() {
  this.unregisterLeaveHook = router.setRouteLeaveHook(route, this.handleRouterWillLeave);
}

componentWillUnmount() {
  if (this.unregisterLeaveHook) {
    this.unregisterLeaveHook();
  }
}
```